docs/dev/coverage.md

```diff
@@ -0,0 +1,173 @@
+# Developer Coverage Overview
+
+## Building
+
+The implementation of coverage starts with the "doCoverage" flag on
+the builder in `comp-builder.nix`. The doCoverage flag enables and
+disables the Cabal coverage flag and copies any generated coverage
+data to "$out/share/hpc".
+
+There also exists a `covered` attribute on each component, it takes
+the existing derivation for that component and modifies it to have
+coverage enabled.
+
+## Mix and tix files
+
+The coverage information for any derivation consists of "mix" and
+"tix" files.
+
+Mix files record static information about a source file and are
+generated at build time. They primarily contain a path to the source
+file and information about expressions and regions of the source file,
+which are later referenced by tix files.
+
+Tix files contain dynamic information about a test run, recording when
+a portion of a source file is touched by a test. These are generated
+when the test is run.
+
+## Coverage reports
+
+### Package reports
+
+The coverage information generated will look something like this:
+
+```bash
+/nix/store/...-my-project-0.1.0.0-coverage-report/
+└── share
+    └── hpc
+        ├── mix
+        │   ├── my-library-0.1.0.0
+        │   │   └── my-library-0.1.0.0-ERSaOroBZhe9awsoBkhmcV
+        │   │       ├── My.Lib.Config.mix
+        │   │       ├── My.Lib.Types.mix
+        │   │       └── My.Lib.Util.mix
+        │   └── my-test-1
+        │       ├── Spec.mix
+        │       └── Main.mix
+        ├── tix
+        │   ├── my-library-0.1.0.0
+        │   │   └── my-library-0.1.0.0.tix
+        │   └── my-test-1
+        │       └── my-test-1.tix
+        └── html
+            ├── my-library-0.1.0.0
+            │   ├── my-library-0.1.0.0-ERSaOroBZhe9awsoBkhmcV
+            │   │   ├── My.Lib.Config.hs.html
+            │   │   ├── My.Lib.Types.hs.html
+            │   │   └── My.Lib.Util.hs.html
+            │   ├── hpc_index_alt.html
+            │   ├── hpc_index_exp.html
+            │   ├── hpc_index_fun.html
+            │   └── hpc_index.html
+            └── my-test-1
+                ├── my-library-0.1.0.0-ERSaOroBZhe9awsoBkhmcV
+                │   ├── My.Lib.Config.hs.html
+                │   ├── My.Lib.Types.hs.html
+                │   └── My.Lib.Util.hs.html
+                ├── hpc_index_alt.html
+                ├── hpc_index_exp.html
+                ├── hpc_index_fun.html
+                └── hpc_index.html
+```
+
+- The mix files are copied verbatim from the builds with coverage.
+- The tix files for each test are copied from the check run verbatim.
+- The tix files for each library are generated by summing the tix
```

Comment:
Is this a standard thing to do? I would have hoped that maybe there were existing tools for working with tix files out there? Maybe we should consider writing one, and then just calling it? I'm a little uncomfortable having sophisticated logic like this just written in bash in a random Nix builder.

Reply:
This applies to the other clever coverage file handling you mention elsewhere.